spectec/src/middlend/sideconditions.ml

```diff
@@ -102,8 +102,23 @@ let rec t_prem env prem = match prem.it with
 
 let t_prems env = List.concat_map (t_prem env)
 
+let rec equiv_exp e1 e2 = Il.Eq.eq_exp e1 e2 ||
+  match e1.it, e2.it with
+  | LenE { it = IterE (_, (ListN ne, _)); _ }, _ -> equiv_exp ne e2
+  | _, LenE { it = IterE (_, (ListN ne, _)); _ } -> equiv_exp e1 ne
+  | _ -> false
```

Comment:
That looks like a useful simplification `(LetE of IterE of ListN)`, independentently of whether it gets optimized.

So maybe better to apply that simplifications in the right spot (e.g. where the `LenE` is introduced, maybe using a smart contructor `lenE` that short-circuits).

Then `equiv_expr` becomes simply `Il.Eq.eq_exp`.

Reply:
Sounds good. I believe using the special constructor in 'same_len' and 'has_len' should be enough.